pages/api/paybutton/download/transactions/[paybuttonId].ts

```diff
@@ -0,0 +1,152 @@
+import Cors from 'cors'
+import moment from 'moment'
+import {
+  PRICE_API_DATE_FORMAT,
+  RESPONSE_MESSAGES,
+  DEFAULT_PAYBUTTON_CSV_FILE_DELIMITER,
+  PAYBUTTON_TRANSACTIONS_FILE_HEADERS,
+  DECIMALS,
+  SUPPORTED_QUOTES,
+  DEFAULT_QUOTE_SLUG
+} from 'constants/index'
+import { TransactionWithAddressAndPrices, fetchTransactionsByPaybuttonId, getTransactionValueInCurrency } from 'services/transactionService'
+import { PaybuttonWithAddresses, fetchPaybuttonById } from 'services/paybuttonService'
+import { streamToCSV } from 'utils/files'
+import { setSession } from 'utils/setSession'
+import { NextApiResponse } from 'next'
+import { Decimal } from '@prisma/client/runtime'
+
+export interface TransactionFileData {
+  amount: Decimal
+  date: moment.Moment
+  value: number
+  rate: number
+  paybuttonName: string
+  transactionId: string
+  currency: string
+}
+
+export interface FormattedTransactionFileData {
+  amount: string
+  date: string
+  value: string
+  rate: string
+  paybuttonName: string
+  transactionId: string
+}
+
+function isCurrencyValid (currency: string): boolean {
+  return SUPPORTED_QUOTES.includes(currency)
+}
+
+const cors = Cors({ methods: ['GET', 'HEAD'] })
+
+const getPaybuttonTransactionsFileData = (transaction: TransactionWithAddressAndPrices, paybutton: PaybuttonWithAddresses, currency: string): TransactionFileData => {
+  const { amount, createdAt, hash } = transaction
+  const value = getTransactionValueInCurrency(transaction, currency)
+  const date = moment(createdAt)
+
+  const rate = value / amount.toNumber()
+
+  return {
+    amount,
+    date,
+    transactionId: hash,
+    value,
+    rate,
+    paybuttonName: paybutton.name,
+    currency
+  }
+}
+
+const formatPaybuttonTransactionsFileData = (data: TransactionFileData): FormattedTransactionFileData => {
+  const {
+    amount,
+    date,
+    value,
+    rate,
+    currency
+  } = data
+  return {
+    ...data,
+    amount: amount.toFixed(DECIMALS[currency]),
+    date: date.format(PRICE_API_DATE_FORMAT),
+    value: value.toFixed(DECIMALS[currency]),
+    rate: rate.toFixed(14)
+  }
+}
+
+export const downloadPaybuttonTransactionsFile = async (
+  res: NextApiResponse,
+  paybutton: PaybuttonWithAddresses,
+  currency: string): Promise<void> => {
+  const transactions = await fetchTransactionsByPaybuttonId(paybutton.id)
+
+  const mappedTransactionsData = transactions.map(tx => {
+    const data = getPaybuttonTransactionsFileData(tx, paybutton, currency)
+    return formatPaybuttonTransactionsFileData(data)
+  })
+  const headers = Object.keys(PAYBUTTON_TRANSACTIONS_FILE_HEADERS)
+  const formmattedHeaders = Object.values(PAYBUTTON_TRANSACTIONS_FILE_HEADERS)
+
+  streamToCSV(
+    mappedTransactionsData,
+    headers,
+    DEFAULT_PAYBUTTON_CSV_FILE_DELIMITER,
+    res,
+    formmattedHeaders
+  )
+}
+
+export default async (req: any, res: any): Promise<void> => {
+  try {
+    await new Promise((resolve, reject) => {
```

Comment:
Why is this needed?

- We only use CORS on endpoints that are supposed to be accessed by the PayButton client (or any external domain, for that matter).
- We have a function called `runMiddleware` in utils to do this set up.